### PR TITLE
Support for sync strategies in the vault

### PR DESCRIPTION
**Pull request by brkomir**
Since we can withdraw funds instantly from synchronous strategies, the code in the vault has been changed to accommodate withdrawing funds when the vault's funds are not enough to cover the transfer by withdrawing the difference from the strategy and rebalancing the vault's funds to maintain the expected minimum funds as reserves.
 
Main changes:
- added support for sync strategies in the vault
- rebalance the vault's funds when calling claim, withdraw, and partialWithdraw
- removed withdrawAllToVault function from the IStrategy interface (simple withdraw is sufficient for this)
- refactored tests to use common helper methods

### CODE REVIEW

File: contracts/strategy/IStrategy.sol [hunk 1]
```diff
@@ -58,23 +58,24 @@ interface IStrategy {
     }
 
     /**
-     * The vault linked to this stragegy
+     * Provides information about wether the strategy is synchronous or asynchronous.
+     * 
+     * @notice Synchronous strategies support instant withdrawals and deposits,
```
Comment (gabrielpoca) [inline contracts/strategy/IStrategy.sol:63]: For the deposits, we don't really change the behaviour

File: contracts/Vault.sol [hunk 2]
```diff
@@ -553,7 +555,7 @@ contract Vault is
         if (_strategy == address(0)) revert VaultStrategyNotSet();
         if (IStrategy(_strategy).vault() != address(this))
             revert VaultInvalidVault();
-        if (address(strategy) != address(0) && strategy.hasAssets() == true)
+        if (address(strategy) != address(0) && strategy.hasAssets())
```
Comment (gabrielpoca) [inline contracts/Vault.sol:558]: why this change?

File: test/VaultAsyncMode.spec.ts [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+
+      await addYieldToVault('50');
+      await vault.connect(owner).updateInvested();
+
+      await expect(
+        vault.connect(alice).claimYield(alice.address),
+      ).to.be.revertedWith('VaultNotEnoughFunds');
+    });
+
+    it("doesn't revert and withdraws funds from the Vault when yield = reserves", async () => {
```
Comment (gabrielpoca) [inline test/VaultAsyncMode.spec.ts:107]: `it("works if the yield = reserves")` ?

File: test/VaultAsyncMode.spec.ts [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+      await expect(vault.connect(alice).claimYield(alice.address)).to.not.be
+        .reverted;
+
+      expect(await underlying.balanceOf(vault.address)).to.eq('0');
+      expect(await underlying.balanceOf(strategy.address)).to.eq(
+        parseUnits('135'),
+      );
+    });
+
+    it("doesn't revert and withdraws funds from the Vault when yield < reserves", async () => {
```
Comment (gabrielpoca) [inline test/VaultAsyncMode.spec.ts:135]: I don't think we need "doesn't revert" section

File: test/VaultAsyncMode.spec.ts [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+  describe("#withdraw doesn't rebalance the Vault's funds", () => {
+    it('reverts when amount > reserves', async () => {
+      const params = depositParams.build({
+        amount: parseUnits('100'),
+        inputToken: underlying.address,
+        claims: [claimParams.percent(100).to(alice.address).build()],
+      });
+      await vault.connect(alice).deposit(params);
+
+      await addYieldToVault('50');
```
Comment (gabrielpoca) [inline test/VaultAsyncMode.spec.ts:174]: why do you need yield on this test and the ones that come after?

File: contracts/Vault.sol (hunk 2 above)
Comment (brkomir) [reply contracts/Vault.sol:558]: it was unnecessary to check if boolean value == true inside an if statement

File: test/VaultAsyncMode.spec.ts (hunk 4 above)
Comment (brkomir) [reply test/VaultAsyncMode.spec.ts:135]: I tried to be more specific but I agree it is not necessary

File: test/VaultAsyncMode.spec.ts (hunk 5 above)
Comment (brkomir) [reply test/VaultAsyncMode.spec.ts:174]: I don't need it actually, going to remove that

File: contracts/strategy/IStrategy.sol (hunk 1 above)
Comment (brkomir) [reply contracts/strategy/IStrategy.sol]: [fixed](https://github.com/lindy-labs/sc_solidity-contracts/pull/136/commits/47be529313b023a0fe04a06247a0095ceba141c9)

File: test/VaultAsyncMode.spec.ts (hunk 3 above)
Comment (brkomir) [reply test/VaultAsyncMode.spec.ts]: [fixed](https://github.com/lindy-labs/sc_solidity-contracts/pull/136/commits/47be529313b023a0fe04a06247a0095ceba141c9)

File: test/VaultAsyncMode.spec.ts (hunk 4 above)
Comment (brkomir) [reply test/VaultAsyncMode.spec.ts]: [fixed](https://github.com/lindy-labs/sc_solidity-contracts/pull/136/commits/47be529313b023a0fe04a06247a0095ceba141c9)

File: test/VaultAsyncMode.spec.ts (hunk 5 above)
Comment (brkomir) [reply test/VaultAsyncMode.spec.ts]: [fixed](https://github.com/lindy-labs/sc_solidity-contracts/pull/136/commits/47be529313b023a0fe04a06247a0095ceba141c9)

**Comment by gabrielpoca** [comment]
@naps62 can you have a look?

File: contracts/mock/MockStrategySync.sol [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+import {AccessControl} from "@openzeppelin/contracts/access/AccessControl.sol";
+
+import {IStrategy} from "../strategy/IStrategy.sol";
+import {CustomErrors} from "../interfaces/CustomErrors.sol";
+
+contract MockStrategySync is IStrategy, AccessControl, CustomErrors {
+    using SafeERC20 for IERC20;
+
+    bytes32 public constant MANAGER_ROLE =
+        0x241ecf16d79d0f8dbfb92cbc07fe17840425976cf0667f022fe9877caa831b08; // keccak256("MANAGER_ROLE");
```
Comment (naps62) [inline contracts/mock/MockStrategySync.sol:15]:
```suggestion
    bytes32 public constant MANAGER_ROLE = keccak256("MANAGER_ROLE");
```

no need to keep the hardocded hash + comment nowadays :)

File: contracts/Vault.sol [hunk 7]
```diff
@@ -647,6 +649,8 @@ contract Vault is
             );
         }
 
+        _rebalanceBeforeTransfer(amount);
+
         underlying.safeTransfer(_to, amount);
```
Comment (naps62) [inline contracts/Vault.sol:654]: sounds like it would be a better pattern to have an internal function covering these two calls, instead of duplicating both of them, everywhere (I'm assuming they're always both together?)

wdyt?

**Review by naps62 — APPROVED** [review]

File: contracts/Vault.sol [hunk 8]
```diff
@@ -674,9 +678,39 @@ contract Vault is
...
+        if (_amount <= vaultBalance) return;
+        if (!strategy.isSync()) revert VaultNotEnoughFunds();
+
+        uint256 expectedReserves = (totalUnderlying() - _amount).pctOf(
+            10000 - investPct
+        );
+
+        // we want to withdraw the from the strategy only what is needed
+        // to cover the transfer and leave the vault with the expected reserves
+        uint256 needed = _amount + expectedReserves - vaultBalance;
```
Comment (naps62) [inline contracts/Vault.sol:707]: another thought that just came to mind. it's arguably very inefficient to always just calculate `needed` and do a withdrawal from the vault. what if `needed = 1`? What if 10 users withdraw from the vault in quick succession? We're incuring on them the gas cost for 10 different rebalances, when we could just have some threshold

e.g.:

```
uint256 desired = _amount + expectedReserves - vaultBalance;
uint256 needed = desired * 90 / 100; // let's say we have a 10% margin of error threshold

// withdraw the original `desired` amount, but only if we don't have enough to cover `needed`, or to cover the withdrawal itself
if (_amount < vaultBalance || needed < vaultBalance) {
  strategy.withdrawToVault(desired);
}
```

@gabrielpoca thoughts?

Comment (gabrielpoca) [reply contracts/Vault.sol:707]: I may be confusing things, but I don't think that scenario exists. The "needed" will always be more than 5% of the total underlying (100% - investment%), because otherwise, this function would not have been called, right? In that scenario where those 10 users withdraw, the one that tries to take more than what's available in the vault would be the only one rebalancing, unless those are big whale deposits that take more than 5% to withdraw. I guess the issue you're pointing out is that one person would be paying for everyone else? Do we pay fees depending on the amount we withdraw?

Comment (naps62) [reply contracts/Vault.sol:707]: Ah, sorry. I forgot when doing this second review that there's a `if (_amount <= vaultBalance) return;` above

File: contracts/Vault.sol (hunk 7 above)
Comment (brkomir) [reply contracts/Vault.sol:654]: This is indeed correct. However, I'm not a fan of functions that imply multiple responsibilities in their name, but in this case, I don't have a strong opinion on that. I'm okay with leaving this as it is right now if that is acceptable to you as well.

File: contracts/Vault.sol [hunk 9]
```diff
@@ -674,9 +678,37 @@ contract Vault is
...
+    /**
+     * Rebalances the vault's funds to cover the transfer of funds from the vault
+     * by disinvesting from the strategy. After the rebalance the vault is left
+     * with a set percentage (100% - invest%) of the total underlying as reserves.
+     *
+     * @notice this will have effect only for sync strategies.
+     *
+     * @param _amount Funds to be transferred from the vault.
+     */
+    function _rebalanceBeforeTransfer(uint256 _amount) internal {
```
[the remark this answers is not in the document]
Comment (brkomir) [reply contracts/Vault.sol]: [fixed](https://github.com/lindy-labs/sc_solidity-contracts/pull/136/commits/159c18a7b58a41190953fd6145799ca29f4bb711)

**Review by gabrielpoca — APPROVED** [review]
